### ISSUE TITLE
Error: No native build was found for platform

### ISSUE THREAD

**Issue by martinszeltins** [vercel/pkg#1660]
### What version of pkg are you using?

5.7.0

### What version of Node.js are you using?

14.19.3

### What operating system are you using?

Ubuntu 20.04

### What CPU architecture are you using?

x86_64

### What Node versions, OSs and CPU architectures are you building for?

node14-linux-x64

### Describe the Bug

For some reason when I use the `serialport` npm package in my app, it compiles but when I try to run it it throws this error (no matter what platform I build for):

````
$ ./app

pkg/prelude/bootstrap.js:1876
      throw error;
      ^

Error: No native build was found for platform=linux arch=x64 runtime=node abi=83 uv=1 libc=glibc node=14.19.2
    loaded from: /snapshot/pkg/node_modules/@serialport/bindings-cpp

    at Function.path (/snapshot/pkg/node_modules/node-gyp-build/index.js:60:9)
    at load (/snapshot/pkg/node_modules/node-gyp-build/index.js:22:30)
    at Object.<anonymous> (/snapshot/pkg/node_modules/@serialport/bindings-cpp/dist/load-bindings.js:10:46)
    at Module._compile (pkg/prelude/bootstrap.js:1930:22)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)
    at Module.load (internal/modules/cjs/loader.js:950:32)
    at Function.Module._load (internal/modules/cjs/loader.js:790:12)
    at Module.require (internal/modules/cjs/loader.js:974:19)
    at Module.require (pkg/prelude/bootstrap.js:1855:31)
    at require (internal/modules/cjs/helpers.js:101:18)
````

Here is my package.json

````
{
  "name": "pkg",
  "version": "1.0.0",
  "dependencies": {
    "serialport": "^10.4.0"
  }
}
````

And here is my app.js

````
const { SerialPort } = require('serialport')

const port = new SerialPort({
  path: '/dev/ttyUSB0',
  baudRate: 19200,
})

console.log('Serial port opened successfully!')
````

And this is how I build:

````
$ pkg app.js --target=node14-linux-x64
> pkg@5.7.0
````

**Comment by patsonkaushik** [comment]
same issue

**Comment by patsonkaushik** [comment]
any solution?

**Comment by martinszeltins** [comment]
> any solution?

@patsonkaushik yes, there is a solution. You need to copy the prebuilt binaries directory to your project root.

**Comment by patsonkaushik** [comment]
> > any solution?
> 
> @patsonkaushik yes, there is a solution. You need to copy the prebuilt binaries directory to your project root.

=>pkg index.js --target linuxstatic

pkg/prelude/bootstrap.js:1876
      throw error;
      ^

Error: No native build was found for platform=linux arch=x64 runtime=node abi=83 uv=1 libc=glibc node=14.19.2
    loaded from: /snapshot/pkg/node_modules/@serialport/bindings-cpp

    at Function.path (/snapshot/pkg/node_modules/node-gyp-build/index.js:60:9)
    at load (/snapshot/pkg/node_modules/node-gyp-build/index.js:22:30)
    at Object.<anonymous> (/snapshot/pkg/node_modules/@serialport/bindings-cpp/dist/load-bindings.js:10:46)
    at Module._compile (pkg/prelude/bootstrap.js:1930:22)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)
    at Module.load (internal/modules/cjs/loader.js:950:32)
    at Function.Module._load (internal/modules/cjs/loader.js:790:12)
    at Module.require (internal/modules/cjs/loader.js:974:19)
    at Module.require (pkg/prelude/bootstrap.js:1855:31)
    at require (internal/modules/cjs/helpers.js:101:18)

**Comment by patsonkaushik** [comment]
@martinszeltins but,not working binaries file example (sqlite3,serialport library)

**Comment by DrSkunk** [comment]
> > any solution?
> 
> @patsonkaushik yes, there is a solution. You need to copy the prebuilt binaries directory to your project root.

Indeed!
TL;DR Copy the folder `node_modules/@serialport/bindings-cpp/prebuilds` to the location of your binary.

So, I was having the same issue and couldn't find the reason. But by looking at the stack trace:
```
Error: No native build was found for platform=linux arch=x64 runtime=node abi=93 uv=1 libc=glibc node=16.15.0
    loaded from: /snapshot/avr-bulk-flasher/node_modules/@serialport/bindings-cpp

    at Function.path (/snapshot/MY_PROJECT/node_modules/node-gyp-build/index.js:61:9)
    at load (/snapshot/MY_PROJECT/node_modules/node-gyp-build/index.js:22:30)
    at Object.<anonymous> (/snapshot/MY_PROJECT/node_modules/@serialport/bindings-cpp/dist/load-bindings.js:10:46)
    at Module._compile (pkg/prelude/bootstrap.js:1930:22)
```

I opened the file referenced here:
```
    at Function.path (/snapshot/MY_PROJECT/node_modules/node-gyp-build/index.js:61:9)
```
which looks like this: (simplified)

```js

load.path = function (dir) {
  dir = path.resolve(dir || '.')
  var nearby = resolve(path.dirname(process.execPath))
  if (nearby) return nearby

  throw new Error('No native build was found for ' + target + '\n    loaded from: ' + dir + '\n')

  function resolve (dir) {
    // Find matching "prebuilds/<platform>-<arch>" directory
    var tuples = readdirSync(path.join(dir, 'prebuilds')).map(parseTuple)
    var tuple = tuples.filter(matchTuple(platform, arch)).sort(compareTuples)[0]
    if (!tuple) return

    // Find most specific flavor first
    var prebuilds = path.join(dir, 'prebuilds', tuple.name)
    var parsed = readdirSync(prebuilds).map(parseTags)
    var candidates = parsed.filter(matchTags(runtime, abi))
    var winner = candidates.sort(compareTags(runtime))[0]
    if (winner) return path.join(prebuilds, winner.file)
  }
}
```

There you can see the pattern where it looks for the native prebuilds.
[Complete source of node-gyp-build](https://github.com/prebuild/node-gyp-build/blob/master/index.js#L25)

**Comment by nganh1** [comment]
> > any solution?
> 
> @patsonkaushik yes, there is a solution. You need to copy the prebuilt binaries directory to your project root.

It doesn't work. Is there any other solution?

I encountered the same problem after the compilation was successful.
```
Error: No native build was found for platform=win32 arch=x64 runtime=node abi=83 uv=1 libc=glibc node=14.20.0
    loaded from: ...\projects\node_modules\usb

    at Function.path (...\project\node_modules\node-gyp-build\index.js:60:9)
    at load (...\project\node_modules\node-gyp-build\index.js:22:30)
    at Object.<anonymous> (...\project\node_modules\usb\dist\usb\bindings.js:8:36)
    at Module._compile (pkg/prelude/bootstrap.js:1930:22)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)
    at Module.load (internal/modules/cjs/loader.js:950:32)
    at Function.Module._load (internal/modules/cjs/loader.js:790:12)
    at Module.require (internal/modules/cjs/loader.js:974:19)
    at Module.require (pkg/prelude/bootstrap.js:1855:31)
    at require (internal/modules/cjs/helpers.js:101:18)
```

**Comment by cliff0412** [comment]
> > any solution?
> 
> @patsonkaushik yes, there is a solution. You need to copy the prebuilt binaries directory to your project root.

prebuild issue solved. however, below error occurs

TypeError: Cannot read properties of undefined (reading 'int64')
    at Object.<anonymous> (/snapshot/california/node_modules/ref-struct-napi/node_modules/ref-napi/lib/ref.js:592:46)
    at Module._compile (pkg/prelude/bootstrap.js:1930:22)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1159:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at Module.require (pkg/prelude/bootstrap.js:1855:31)
    at require (node:internal/modules/cjs/helpers:102:18)
    at Object.<anonymous> (/snapshot/california/node_modules/ref-struct-napi/lib/struct.js:49:11)
    at Module._compile (pkg/prelude/bootstrap.js:1930:22)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1159:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at Module.require (pkg/prelude/bootstrap.js:1855:31)
    at require (node:internal/modules/cjs/helpers:102:18)

**Comment by nganh1** [comment]
I added the corresponding module to the "assets" to solve this problem.

package.json
```
{
  ...
  "pkg": {
    "outputPath": ".dist",
    "targets": [
      "node14-win-x64",
      "node14-macos-x64",
      "node14-linux-x64"
    ],
    "assets": [
      "utils/*",
      "public/*",
      "public/**/*",
      "node_modules/usb/*",
      "node_modules/usb/**/*"
    ]
  },
  ...
}
```

**Comment by ssk7833** [comment]
> I added the corresponding module to the "assets" to solve this problem.

I added `node_modules/@serialport/bindings-cpp/prebuilds` to the assets then problem solved.

package.json
 
```
{
  ...
  "pkg": {
    "outputPath": ".dist",
    "targets": [
      ...
    ],
    "assets": [
      ...,
      "node_modules/@serialport/bindings-cpp/prebuilds"
    ]
  },
  ...
}
```

**Comment by geoff882** [comment]
> > any solution?
> 
> @patsonkaushik yes, there is a solution. You need to copy the prebuilt binaries directory to your project root.

Sorry, what exactly do you mean binary directories?    

project/
project/dist
project/node_modules/usb

And the files in node_modules/usb/prebuilds
android-arm
android-arm64
darwin-x64+arm64
linux-arm
linux-arm64
linux-ia32
linux-x64
win32-ia32
win32-x64

Tried all the solutions above.  My error:
Error: No native build was found for platform=darwin arch=arm64 runtime=node abi=93 uv=1 armv=8 libc=glibc node=16.16.0
    loaded from: /snapshot/src/printserver/node_modules/usb

    at Function.path (/snapshot/src/printserver/node_modules/node-gyp-build/index.js:60:9)

**Comment by heraclesly** [comment]
> 

Hello,did you solve this problem?I got the same Error,it's happen when i import serialport module in my main.js

**Comment by heraclesly** [comment]
> > > any solution?
> > 
> > 
> > @patsonkaushik yes, there is a solution. You need to copy the prebuilt binaries directory to your project root.
> 
> Sorry, what exactly do you mean binary directories?
> 
> project/ project/dist project/node_modules/usb
> 
> And the files in node_modules/usb/prebuilds android-arm android-arm64 darwin-x64+arm64 linux-arm linux-arm64 linux-ia32 linux-x64 win32-ia32 win32-x64
> 
> Tried all the solutions above. My error: Error: No native build was found for platform=darwin arch=arm64 runtime=node abi=93 uv=1 armv=8 libc=glibc node=16.16.0 loaded from: /snapshot/src/printserver/node_modules/usb
> 
> ```
> at Function.path (/snapshot/src/printserver/node_modules/node-gyp-build/index.js:60:9)
> ```


Hello,did you solve this problem?I got the same Error,it's happen when i import serialport module in my main.js

**Comment by Didi1913** [comment]
Anyone with the solution do you mind describing the solution in more detail?

**Comment by neko-para** [comment]
Here's my solution.
I'm use ffi-napi and ref-napi, which contains native module.
Firstly I use esbuild to pack my project, while mark these two module as external
```shell
esbuild --platform=node --bundle packages/server/src/index.ts --outfile=dist/index.js --external:@tigerconnect/ref-napi --external:@tigerconnect/ffi-napi
```
Then include these two module into assets
```json
{
  "bin": "dist/index.js",
  "pkg": {
    "assets": [
      "node_modules/@tigerconnect/ffi-napi/**/*",
      "node_modules/@tigerconnect/ref-napi/**/*"
    ]
  },
}
```

**Comment by javierguzman** [comment]
Could someone explain my why this happens with that package? I am facing this issue without pkg and I would really understand what it is going on

**Comment by neko-para** [comment]
> Could someone explain my why this happens with that package? I am facing this issue without pkg and I would really understand what it is going on

Those packages contains prebuilt binaries and will be loaded dynamicly, which is hard for pack tools to keep them working. Actually I'm not sure if keeping binaries could help, but I choose to treat them as external packages and keep the directory structures (as assets instead of scripts)

**Comment by javierguzman** [comment]
Thank you very much @neko-para ! So it is like those packages are built already for my machine and they are not pure JS so the bundlers like Webpack get confused somehow, would that be? Then the solution would be to tell Webpack ok, ignore those guys and then move those binaries/built files directly into the webpack output built folder. Would that be correct? I have been really struggling with this error using Electron :( that is why I am keen on understanding well the details !

**Comment by stenya** [comment]
@javierguzman Did you manage it with Webpack? I would be grateful for any info. 
I am facing the same problem: `Cannot read properties of undefined (reading 'int64')`.

I use Webpack + Vue + ElectronBuilder.

**Comment by robertsLando** [comment]
For anyone that is interested in this project, giving that I don't receive any feedback from Vercel anymore, I created my own forks of both pkg and pkg-fetch:

[yao-pkg/pkg-fetch](https://github.com/yao-pkg/pkg-fetch)
[yao-pkg/pkg](https://github.com/yao-pkg/pkg)

If someone is interested in helping me create a PR or move your active PR to those repositories.

Thanks 🙏🏼

**Comment by javierguzman** [comment]
@stenya I think I did but I do not remember much of it and I am not in the same company I was back then so no access to repo

**Comment by neko-para** [comment]
> @javierguzman Did you manage it with Webpack? I would be grateful for any info. I am facing the same problem: `Cannot read properties of undefined (reading 'int64')`.
> 
> I use Webpack + Vue + ElectronBuilder.

If you just want any ffi work with webpack, you can take a look at [koffi](http://koffi.dev/), at least this package provides official examples for using with electron-forge/webpack.

**Comment by stenya** [comment]
> @javierguzman Did you manage it with Webpack? I would be grateful for any info. I am facing the same problem: `Cannot read properties of undefined (reading 'int64')`.
> 
> I use Webpack + Vue + ElectronBuilder.

If anyone is interested, here is the solution for ElectronBuilder which works for me: https://github.com/node-ffi-napi/node-ffi-napi/issues/245
